pandas/core/arrays/sparse.py

```diff
@@ -1215,10 +1215,26 @@ def astype(self, dtype=None, copy=True):
         dtype = pandas_dtype(dtype)
 
         if not isinstance(dtype, SparseDtype):
-            dtype = SparseDtype(dtype, fill_value=self.fill_value)
+            fill_value = astype_nansafe(np.array(self.fill_value),
+                                        dtype).item()
+            dtype = SparseDtype(dtype, fill_value=fill_value)
+
+        # Typically we'll just astype the sp_values to dtype.subtype,
```

Comment:
can you make this a method on the Dtype itself to avoid cluttering this up here? maybe ``dtype.astype_type``
alternatively . we could actually add ``.astype_nansafe(value, copy=False)`` as a Dtype method (kind of makes sense actually)

Reply:
I was playing around with that earlier (didn't push it though). I called it `SparseDtype.astype`. I'll give it another shot and see what it de-duplicates.

Reply:
Actually, can you clarify what you had in mind for `astype_nansafe(value, copy=False)`? What would value here be? An array or a scalar?

I'll have a followup PR soon (hopefully today) for ensuring that the dtype of `SparseArray.sp_values` is consistent with the type of `SparseArray.dtype.fill_value`. I think my `SparseDtype.astype` is more useful there. It wouldn't make sense for using here, since we're astyping the actual array of values.

Reply:
Updated to add two methods

1. `SparseDtype.astype`: convert from a SparseDtype to a new dtype, taking care to astype `self.fill_value` if needed.
2. `SparseDtype._subtype_with_str` to hold the logic for determining what the "real" subtype is, if we actually want str.